core/state_transition.go

```diff
@@ -314,29 +394,35 @@ func GetGasBalancesInBig(state vm.StateDB, chainconfig *params.ChainConfig, acco
 	return bal.ToBig(), sgtBal.ToBig()
 }
 
-func (st *StateTransition) SubSoulBalance(account common.Address, amount *big.Int, reason tracing.BalanceChangeReason) (err error) {
-	current := st.GetSoulBalance(account).ToBig()
+// called by buyGas
+func (st *StateTransition) subSoulBalance(account common.Address, amount *uint256.Int, reason tracing.BalanceChangeReason) (err error) {
+	current := st.GetSoulBalance(account)
 	if current.Cmp(amount) < 0 {
 		return fmt.Errorf("soul balance not enough, current:%v, expect:%v", current, amount)
 	}
 
-	value := uint256.MustFromBig(current.Sub(current, amount)).Bytes32()
+	value := current.Sub(current, amount).Bytes32()
 	st.state.SetState(types.SoulGasTokenAddr, TargetSGTBalanceSlot(account), value)
 
 	if st.evm.ChainConfig().IsOptimism() && st.evm.ChainConfig().Optimism.IsSoulBackedByNative {
-		st.state.SubBalance(types.SoulGasTokenAddr, uint256.MustFromBig(amount), reason)
+		st.state.SubBalance(types.SoulGasTokenAddr, amount, reason)
 	}
+
+	// subSoulBalance is only called by buyGas once
+	st.usedSGTBalance = amount
```

Comment:
I feel this line of code should be moved out of this function so that the usage of `subXX` and `st.balance=xxx` is aligned between native balance and SGT balance.

Reply:
Done in d5c1a66ca5e7e0524bedb75d9188a14a5a8eee21.